Hotfix/db stream next

ricdezen opened the pull request:
## Premise
Fixed some issues with the db stream.
It cannot directly subclass the Stream class due to inability to modify the cursor by index, but now behaves correctly as an Unmodifiable Stream.

## Features
**fix (profile_batch.py):** changed to __next__ calls instead of for loop
**fix (database_stream.py):**
**+** changed DatabaseStream and Iterator to better match standard Stream structure
**+** Added internal buffer to PostgreSQLIterator for better has_next call
**+** Added parent Stream reference to iterator, to allow closing the parent when the Stream is over.

## Picture of a cute animal
You MUST place a pic or a gif of a cute animal at the end of the PR.
![Cute](https://media2.giphy.com/media/5i7umUqAOYYEw/giphy.gif?cid=ecf05e478fddbba1ddf9f9cceb68f655c8eee989efe381a3&rid=giphy.gif)

Commit:
fix (database_stream.py): added missing return statement

Commit:
Merge branch 'feat/database-stream' of https://github.com/OTRI-Unipd/OTRI into feat/database-stream

Commit:
Merge branch 'feat/filtering' of https://github.com/OTRI-Unipd/OTRI into feat/database-stream

Commit:
fix (profile_batch.py): changed to __next__ calls instead of for loop

fix (database_stream.py): changed DatabaseStream and Iterator to better match standard Stream structure
+ Added internal buffer to PostgreSQL iterator for better has_next calls

Commit:
feat (database_stream.py) :
+ added reference to parent strem to iterator, so that it can be closed when the stream is over
+ added check for cursor status on __next__ iterator method

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -52,25 +78,38 @@ def __next__(self):
                 When no result is given by the database (the query is over),
```
Update this specification, it's outdated

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -52,25 +78,38 @@ def __next__(self):
     def has_next(self) -> bool:
         '''
         Returns:
             True if the cursor is still open (there are still rows to retrieve)
```
True if there's another item, false otherwise. Let's remove the dependency from the cursor to the outside environment.

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -81,17 +120,22 @@ class PostgreSQLStream(DatabaseStream):
-        self.__query = query
         self.__batch_size = batch_size
+        self.__cursor = self.__new_cursor(connection)
+        self.__cursor.execute("SELECT data_json as json FROM {} WHERE {};".format(
```
I was thinking about the same solution: moving the query in the constructor.

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -100,18 +144,36 @@ def __iter__(self):
             psycopg2.errors.* :
                 if the query is not correct due to syntax or wrong names.
```
Doesn't raise exceptions anymore

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -52,25 +78,38 @@ def __next__(self):
 
     def close(self):
         '''
         Closes the cursor.
```
And the parent stream

CremaLuca reviewed (changes requested):
Some minor documentation things you missed in the hurry.
The code sounds like it should run properly in the filters.

CremaLuca commented on otri/database/database_stream.py:
```diff
@@ -29,19 +33,41 @@ def __iter__(self) -> DatabaseIterator:
+        self.__parent = parent
+        self.__buffer = None
 
     def __next__(self):
```
I noticed that the functions returns a tuple containing a single atom and a comma "({...},)", is it reasonable to return `next(self.__cursor)[0]` and `item[0]`? They're probably going to be always atoms I think.

ricdezen replied:
Maybe, but maybe we might want to return their id in the database to modify that row, so I think it would be best to return the whole row, but this way we need a filter that turns each row into a json dict

Commit:
fix (database_stream.py): various minor specs updates

Commit:
fix (database_stream.py) : OCD

Commit:
fix (nuplicator_filter.py) : made unnecessarily public fields into private

Commit:
feat (generic_filter.py):
+ Added class GenericFilter that applies some Callable to a Stream

test (generic_filter_test.py) : basic tests
